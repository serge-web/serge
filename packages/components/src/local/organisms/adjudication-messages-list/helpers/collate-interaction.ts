
import { Asset, ForceData, GroupedActivitySet, MessageInteraction, MessagePlanning, PerForcePlanningActivitySet, PlanningActivity, PlanningMessageStructure } from '@serge/custom-types'
import { findAsset, ForceStyle } from '@serge/helpers'
import _ from 'lodash'

export type InteractionData = {
  interaction: MessageInteraction
  order1: MessagePlanning
  order2: MessagePlanning | undefined
  allAssets: Asset[]
  allAssetNames: string[]
  order1Activity: PlanningActivity | undefined
  order2Activity: PlanningActivity | undefined
}

const getActivity = (activities:PerForcePlanningActivitySet[], forceId?: ForceData['uniqid'], activityId?: PlanningMessageStructure['activity']): PlanningActivity | undefined => {
  const force = activities.find((act) => act.force === forceId)
  if (!force) {
    throw Error('Failed to find group for:' + force)
  }
  const groupAct = force.groupedActivities.find((group: GroupedActivitySet) =>
    group.activities.find((act: string | PlanningActivity) => {
      if (typeof act === 'string') {
        throw Error('Activity should not be string')
      }
      const actObj = act as PlanningActivity
      return actObj.uniqid === activityId
    })
  )
  if (!groupAct) {
    throw Error('failed to find group activity')
  }

  const activity = groupAct.activities.find((act: string | PlanningActivity) => {
    if (typeof act === 'string') {
      throw Error('Activity should not be string')
    }
    const actObj = act as PlanningActivity
    return actObj.uniqid === activityId
  }) as PlanningActivity
  if (!activity) {
    throw Error('failed to find activity')
  }
  return activity
}

export const updateAssets = (asset: Record<string, any>, interaction: InteractionData): Record<string, any> => {
  if (asset !== undefined) {
    const allAssIds = interaction.allAssets.map((asset) => asset.uniqid)
    const allAssNames = interaction.allAssets.map((asset) => asset.name)
    asset.enum = allAssIds
    asset.options.enum_titles = allAssNames
  }
  return asset
}

export const updateForces = (force: Record<string, any>, forces: ForceStyle[]): Record<string, any> => {
  if (force !== undefined) {
    const allForceIds = forces.map((force) => force.forceId)
    const allForceNames = forces.map((force) => force.force)
    force.enum = allForceIds
    force.options.enum_titles = allForceNames
  }
  return force
}

export const collateInteraction = (intId: string, interactionMessages: MessageInteraction[],
  planningMessages: MessagePlanning[], forces: ForceData[], forceStyles: ForceStyle[], forcePlanningActivities?: PerForcePlanningActivitySet[]): InteractionData => {
  const intMsg = interactionMessages.find((value) => value._id === intId)
  if (!intMsg) {
    throw Error('Failed to find interaction message:' + intId)
  }
  const interaction = intMsg.details.interaction
  if (!interaction) {
    throw Error('Failed to find interaction message:' + intId)
  }
  const order1AssetsIds: string[] = []
  const order1 = planningMessages.find((plan) => plan._id === interaction.orders1)
  if (!order1) {
    throw Error('Failed to find interaction message:' + interaction.orders1)
  }
  order1.message.ownAssets && order1AssetsIds.push(...order1.message.ownAssets.map((asset) => asset.asset))
  order1.message.otherAssets && order1AssetsIds.push(...order1.message.otherAssets)

  const order2AssetsIds: string[] = []
  const order2: MessagePlanning | undefined = interaction.orders2 ? planningMessages.find((plan): boolean => plan._id === interaction.orders2) : undefined
  if (order2) {
    order2.message.ownAssets && order2AssetsIds.push(...order2.message.ownAssets.map((asset) => asset.asset))
    order2.message.otherAssets && order2AssetsIds.push(...order2.message.otherAssets)
  }

  const allIds = order1AssetsIds.concat(order2AssetsIds)
  const uniqIds = _.uniq(allIds)
  const allAssets = uniqIds.map((id: string) => findAsset(forces, id))
  const sortedAllAssets = _.sortBy(allAssets, (a: Asset) => a.name)
  const sortedAllAssetNames = sortedAllAssets.map((asset: Asset) => {
    const force = order1AssetsIds.includes(asset.uniqid)
      ? forceStyles.find((force) => force.forceId === order1.details.from.forceId)
      : forceStyles.find((force) => force.forceId === (order2 && order2.details.from.forceId))
    if (!force) {
      throw Error('Failed to find force for:' + asset.uniqid)
    }
    return force.force + '-' + asset.name
  })

  const act1 = forcePlanningActivities && getActivity(forcePlanningActivities, order1.details.from.forceId, order1.message.activity)
  const act2 = order2 && forcePlanningActivities && getActivity(forcePlanningActivities, order1.details.from.forceId, order1.message.activity)

  return {
    interaction: intMsg,
    order1: order1,
    order2: order2,
    allAssets: sortedAllAssets,
    allAssetNames: sortedAllAssetNames,
    order1Activity: act1,
    order2Activity: act2
  }
}
