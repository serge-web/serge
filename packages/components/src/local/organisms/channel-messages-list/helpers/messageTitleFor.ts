import { MessageCustom } from '@serge/custom-types'
export default function messageTitleFor (message: MessageCustom): string {
  if (message.message.title || message.message.Title || message.details.messageType) {
    return message.message.title || message.message.Title || message.details.messageType
  } else if (message.message.content) {
    // yes, we have content (probably chat) use it
    return message.message.content
  } else {
    // no content, just use message-type
    return message.details.messageType
  }
}
