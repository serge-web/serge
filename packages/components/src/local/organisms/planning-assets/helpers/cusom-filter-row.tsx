import { Column, MTableFilterRow } from '@material-table/core'
import { Asset, ForceData } from '@serge/custom-types'
import React, { useState } from 'react'

type CustomFilterRowProps = {
  columns: Column<any>[]
  onFilterChanged: (columnId: number, value: string[]) => void
  forces: ForceData[]
}

const CustomFilterRow: React.FC<CustomFilterRowProps> = (props): React.ReactElement => {
  const [localProps, setLocalProps] = useState<CustomFilterRowProps>(props)

  const onFilterChanged = (columnId: number, filter: string[]) => {
    props.onFilterChanged(columnId, filter)
    if (columnId !== 1) {
      return
    }
    const filterSubTypes = {}
    props.forces.forEach(force => {
      if (!force.assets) {
        return
      }
      force.assets.forEach((asset: Asset) => {
        if (asset.attributes && filter.includes(asset.platformTypeId)) {
          const subType = asset.attributes.a_Type as string
          filterSubTypes[subType] = subType
        }
      })
      const subTypeColIdx = props.columns.findIndex(col => col.field === 'subType')
      if (props.columns[subTypeColIdx].lookup) {
        props.columns[subTypeColIdx].lookup = filterSubTypes
      }
      setLocalProps({ ...props })
    })
  }

  return <MTableFilterRow {...localProps} onFilterChanged={onFilterChanged}></MTableFilterRow>
}

export default CustomFilterRow
