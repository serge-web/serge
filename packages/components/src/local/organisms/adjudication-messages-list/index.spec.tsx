/* global it expect */
import { ChannelPlanning, MessageInteraction } from '@serge/custom-types'
import moment from 'moment-timezone'
import React from 'react'
import renderer from 'react-test-renderer'
import { AdjudicationMessagesList } from './index'

import { forceColors } from '@serge/helpers'
import { P9Mock, planningMessageTemplatesMock } from '@serge/mocks'
import { noop } from 'lodash'

const planningChannel = P9Mock.data.channels.channels[0] as ChannelPlanning
const forces = P9Mock.data.forces.forces
const blueForce = forces[1]
const blueRole = blueForce.roles[0]

describe('AdjudicationMessagesList component: ', () => {
  it('renders component correctly', () => {
    moment.tz.setDefault('Etc/UTC')
    const markAllAsRead = (): void => window.alert('Callback on mark all as read')
    const messages: MessageInteraction[] = []

    const tree = renderer
      .create(<AdjudicationMessagesList selectedOrders={[]} setSelectedOrders={noop} forces={forces} template={planningMessageTemplatesMock[0]} gameDate={P9Mock.data.overview.gameDate} channel={planningChannel} hideForcesInChannel={false}
        messages={messages} onRead={undefined} forceColors={forceColors(forces)} onUnread={undefined} isUmpire={true} playerRoleId={blueRole.roleId}
        playerForceId={blueForce.uniqid} onMarkAllAsRead={markAllAsRead} />)
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})
