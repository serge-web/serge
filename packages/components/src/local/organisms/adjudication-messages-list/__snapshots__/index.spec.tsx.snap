// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AdjudicationMessagesList component:  renders component correctly 1`] = `
<div
  className="messages-list"
>
  <div
    className="button-wrap"
  >
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSecondary"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiButton-label"
      >
        <span
          className="MuiButton-startIcon MuiButton-iconSizeMedium"
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M17 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V7l-4-4zm-5 16c-1.66 0-3-1.34-3-3s1.34-3 3-3 3 1.34 3 3-1.34 3-3 3zm3-10H5V5h10v4z"
            />
          </svg>
        </span>
        Get Next
      </span>
      <span
        className="MuiTouchRipple-root"
      />
    </button>
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSecondary"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiButton-label"
      >
        <span
          className="MuiButton-startIcon MuiButton-iconSizeMedium"
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
            />
          </svg>
        </span>
        Create manual
      </span>
      <span
        className="MuiTouchRipple-root"
      />
    </button>
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSecondary"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiButton-label"
      >
        <span
          className="MuiButton-startIcon MuiButton-iconSizeMedium"
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M18 4H6v2l6.5 6L6 18v2h12v-3h-7l5-5-5-5h7z"
            />
          </svg>
        </span>
        # Remaining
      </span>
      <span
        className="MuiTouchRipple-root"
      />
    </button>
    <div
      className="MuiChip-root"
      onKeyDown={[Function]}
      onKeyUp={[Function]}
    >
      <span
        className="MuiChip-label"
      >
        pending
      </span>
    </div>
  </div>
  <div
    className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation2 css-1mbunh-MuiPaper-root"
    style={
      Object {
        "position": "relative",
      }
    }
  >
    <div
      className="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-ig9rso-MuiToolbar-root"
    >
      <div
        className="MuiBox-root css-i6bazn"
      >
        <h6
          className="MuiTypography-root MuiTypography-h6 css-2ulfj5-MuiTypography-root"
          style={
            Object {
              "overflow": "hidden",
              "textOverflow": "ellipsis",
              "whiteSpace": "nowrap",
            }
          }
        >
          Adjudication
        </h6>
      </div>
      <div
        className="MuiBox-root css-15k1nh5"
      />
      <div
        className="MuiFormControl-root MuiTextField-root css-28z0on-MuiFormControl-root-MuiTextField-root"
      >
        <div
          className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-adornedStart MuiInputBase-adornedEnd css-whebh7-MuiInputBase-root-MuiInput-root"
          onClick={[Function]}
          style={Object {}}
        >
          <div
            className="MuiInputAdornment-root MuiInputAdornment-positionStart MuiInputAdornment-standard MuiInputAdornment-sizeMedium css-ittuaa-MuiInputAdornment-root"
          >
            <span
              className="notranslate"
            >
              â€‹
            </span>
            <svg
              aria-hidden={true}
              aria-label="Search"
              aria-labelledby={null}
              className="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall"
              data-mui-internal-clone-element={true}
              focusable="false"
              onBlur={[Function]}
              onFocus={[Function]}
              onMouseLeave={[Function]}
              onMouseOver={[Function]}
              onTouchEnd={[Function]}
              onTouchStart={[Function]}
              viewBox="0 0 24 24"
            >
              <path
                d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
              />
            </svg>
          </div>
          <input
            aria-invalid={false}
            aria-label="Search"
            autoFocus={false}
            className="MuiInputBase-input MuiInput-input MuiInputBase-inputAdornedStart MuiInputBase-inputAdornedEnd css-1x51dt5-MuiInputBase-input-MuiInput-input"
            disabled={false}
            id="mui-2"
            onAnimationStart={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            placeholder="Search"
            required={false}
            type="text"
            value=""
          />
          <div
            className="MuiInputAdornment-root MuiInputAdornment-positionEnd MuiInputAdornment-standard MuiInputAdornment-sizeMedium css-1laqsz7-MuiInputAdornment-root"
          >
            <button
              aria-label="Clear Search"
              className="MuiButtonBase-root Mui-disabled MuiIconButton-root Mui-disabled MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
              disabled={true}
              onBlur={[Function]}
              onClick={[Function]}
              onContextMenu={[Function]}
              onDragLeave={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              tabIndex={-1}
              type="button"
            >
              <svg
                aria-hidden={true}
                aria-label="clear"
                className="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                />
              </svg>
            </button>
          </div>
        </div>
      </div>
      <div
        className="MuiBox-root css-p9qzma"
      >
        <div>
          <div
            style={
              Object {
                "display": "flex",
              }
            }
          >
            <span>
              <button
                aria-label="Show Columns"
                aria-labelledby={null}
                className="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-sizeMedium css-zylse7-MuiButtonBase-root-MuiIconButton-root"
                data-mui-internal-clone-element={true}
                disabled={false}
                onBlur={[Function]}
                onClick={[Function]}
                onContextMenu={[Function]}
                onDragLeave={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                onMouseDown={[Function]}
                onMouseLeave={[Function]}
                onMouseOver={[Function]}
                onMouseUp={[Function]}
                onTouchEnd={[Function]}
                onTouchMove={[Function]}
                onTouchStart={[Function]}
                tabIndex={0}
                type="button"
              >
                <svg
                  aria-hidden={true}
                  className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="view_column"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M14.67 5v14H9.33V5h5.34zm1 14H21V5h-5.33v14zm-7.34 0V5H3v14h5.33z"
                  />
                </svg>
                <span
                  className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </button>
            </span>
            <span>
              <div
                style={
                  Object {
                    "display": "flex",
                  }
                }
              >
                <button
                  aria-label="Only show open interactions"
                  aria-labelledby={null}
                  className="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-sizeMedium css-zylse7-MuiButtonBase-root-MuiIconButton-root"
                  data-mui-internal-clone-element={true}
                  disabled={false}
                  onBlur={[Function]}
                  onClick={[Function]}
                  onContextMenu={[Function]}
                  onDragLeave={[Function]}
                  onFocus={[Function]}
                  onKeyDown={[Function]}
                  onKeyUp={[Function]}
                  onMouseDown={[Function]}
                  onMouseLeave={[Function]}
                  onMouseOver={[Function]}
                  onMouseUp={[Function]}
                  onTouchEnd={[Function]}
                  onTouchMove={[Function]}
                  onTouchStart={[Function]}
                  tabIndex={0}
                  type="button"
                >
                  <svg
                    className="svg-inline--fa fa-user"
                  />
                  <span
                    className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                  />
                </button>
                <button
                  aria-label="Check orders valid"
                  aria-labelledby={null}
                  className="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-sizeMedium css-zylse7-MuiButtonBase-root-MuiIconButton-root"
                  data-mui-internal-clone-element={true}
                  disabled={false}
                  onBlur={[Function]}
                  onClick={[Function]}
                  onContextMenu={[Function]}
                  onDragLeave={[Function]}
                  onFocus={[Function]}
                  onKeyDown={[Function]}
                  onKeyUp={[Function]}
                  onMouseDown={[Function]}
                  onMouseLeave={[Function]}
                  onMouseOver={[Function]}
                  onMouseUp={[Function]}
                  onTouchEnd={[Function]}
                  onTouchMove={[Function]}
                  onTouchStart={[Function]}
                  tabIndex={0}
                  type="button"
                >
                  <svg
                    className="svg-inline--fa fa-bug"
                  />
                  <span
                    className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                  />
                </button>
                <button
                  aria-label="Show filter controls"
                  aria-labelledby={null}
                  className="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-sizeMedium css-zylse7-MuiButtonBase-root-MuiIconButton-root"
                  data-mui-internal-clone-element={true}
                  disabled={false}
                  onBlur={[Function]}
                  onClick={[Function]}
                  onContextMenu={[Function]}
                  onDragLeave={[Function]}
                  onFocus={[Function]}
                  onKeyDown={[Function]}
                  onKeyUp={[Function]}
                  onMouseDown={[Function]}
                  onMouseLeave={[Function]}
                  onMouseOver={[Function]}
                  onMouseUp={[Function]}
                  onTouchEnd={[Function]}
                  onTouchMove={[Function]}
                  onTouchStart={[Function]}
                  tabIndex={0}
                  type="button"
                >
                  <svg
                    className="svg-inline--fa fa-magnifying-glass-plus"
                  />
                  <span
                    className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                  />
                </button>
              </div>
            </span>
          </div>
        </div>
      </div>
    </div>
    <div
      className="MuiBox-root css-187halg"
    >
      <div>
        <div
          style={
            Object {
              "maxHeight": undefined,
              "minHeight": undefined,
              "overflowY": "auto",
            }
          }
        >
          <div>
            <table
              className="MuiTable-root css-rqglhn-MuiTable-root"
              role={null}
              style={
                Object {
                  "tableLayout": "auto",
                }
              }
            >
              <thead
                className="MuiTableHead-root css-15wwp11-MuiTableHead-root"
                role={null}
              >
                <tr
                  className="MuiTableRow-root MuiTableRow-head css-16m31cp-MuiTableRow-root"
                  role={null}
                >
                  <th
                    aria-sort={null}
                    className="MuiTableCell-root MuiTableCell-head MuiTableCell-paddingNone MuiTableCell-sizeMedium css-2ug12u-MuiTableCell-root"
                    scope="col"
                  />
                </tr>
              </thead>
              <tbody
                className="MuiTableBody-root css-apqrd9-MuiTableBody-root"
                role={null}
              >
                <tr
                  className="MuiTableRow-root css-1q1u3t4-MuiTableRow-root"
                  role={null}
                  style={
                    Object {
                      "height": 49,
                    }
                  }
                >
                  <td
                    aria-sort={null}
                    className="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
                    colSpan={1}
                    style={
                      Object {
                        "paddingBottom": 0,
                        "paddingTop": 0,
                        "textAlign": "center",
                      }
                    }
                  >
                    No records to display
                  </td>
                </tr>
              </tbody>
            </table>
          </div>
        </div>
      </div>
    </div>
    <table
      className="MuiTable-root css-rqglhn-MuiTable-root"
      role={null}
    >
      <tfoot
        className="MuiTableFooter-root css-9cgc89-MuiTableFooter-root"
        role={null}
        style={
          Object {
            "display": "grid",
          }
        }
      >
        <tr
          className="MuiTableRow-root MuiTableRow-footer css-1q1u3t4-MuiTableRow-root"
          role={null}
          style={
            Object {
              "display": "grid",
            }
          }
        >
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-footer MuiTableCell-sizeMedium MuiTablePagination-root css-d63lsq-MuiTableCell-root-MuiTablePagination-root"
            colSpan={3}
          >
            <div
              className="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular MuiTablePagination-toolbar css-78c6dr-MuiToolbar-root-MuiTablePagination-toolbar"
            >
              <div
                className="MuiTablePagination-spacer css-1psng7p-MuiTablePagination-spacer"
              />
              <p
                className="MuiTablePagination-selectLabel css-pdct74-MuiTablePagination-selectLabel"
                id="mui-12"
              >
                Rows per page:
              </p>
              <div
                className="MuiInputBase-root MuiInputBase-colorPrimary css-16c50h-MuiInputBase-root-MuiTablePagination-select"
                onClick={[Function]}
              >
                <div
                  aria-expanded="false"
                  aria-haspopup="listbox"
                  aria-labelledby="mui-12 mui-11"
                  className="MuiSelect-select MuiTablePagination-select MuiSelect-standard MuiInputBase-input css-194a1fa-MuiSelect-select-MuiInputBase-input"
                  id="mui-11"
                  onBlur={[Function]}
                  onFocus={[Function]}
                  onKeyDown={[Function]}
                  onMouseDown={[Function]}
                  role="button"
                  tabIndex={0}
                >
                  <div
                    className="MuiBox-root css-c8mw91"
                  >
                    20 rows 
                  </div>
                </div>
                <input
                  aria-hidden={true}
                  className="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
                  onAnimationStart={[Function]}
                  onChange={[Function]}
                  tabIndex={-1}
                  value={20}
                />
                <svg
                  aria-hidden={true}
                  className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiTablePagination-selectIcon MuiSelect-iconStandard css-pqjvzy-MuiSvgIcon-root-MuiSelect-icon"
                  data-testid="ArrowDropDownIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M7 10l5 5 5-5z"
                  />
                </svg>
              </div>
              <p
                className="MuiTablePagination-displayedRows css-levciy-MuiTablePagination-displayedRows"
              >
                0â€“0 of 0
              </p>
              <div
                className="MuiBox-root css-lm78jv"
              >
                <span
                  aria-label="First Page"
                  aria-labelledby={null}
                  className=""
                  data-mui-internal-clone-element={true}
                  onBlur={[Function]}
                  onFocus={[Function]}
                  onMouseLeave={[Function]}
                  onMouseOver={[Function]}
                  onTouchEnd={[Function]}
                  onTouchStart={[Function]}
                >
                  <button
                    aria-label="First Page"
                    className="MuiButtonBase-root Mui-disabled MuiIconButton-root Mui-disabled MuiIconButton-sizeLarge css-mf1cb5-MuiButtonBase-root-MuiIconButton-root"
                    disabled={true}
                    onBlur={[Function]}
                    onClick={[Function]}
                    onContextMenu={[Function]}
                    onDragLeave={[Function]}
                    onFocus={[Function]}
                    onKeyDown={[Function]}
                    onKeyUp={[Function]}
                    onMouseDown={[Function]}
                    onMouseLeave={[Function]}
                    onMouseUp={[Function]}
                    onTouchEnd={[Function]}
                    onTouchMove={[Function]}
                    onTouchStart={[Function]}
                    tabIndex={-1}
                    type="button"
                  >
                    <svg
                      aria-hidden={true}
                      className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                      data-testid="first_page"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M18.41 16.59 13.82 12l4.59-4.59L17 6l-6 6 6 6zM6 6h2v12H6z"
                      />
                    </svg>
                  </button>
                </span>
                <span
                  aria-label="Previous Page"
                  aria-labelledby={null}
                  className=""
                  data-mui-internal-clone-element={true}
                  onBlur={[Function]}
                  onFocus={[Function]}
                  onMouseLeave={[Function]}
                  onMouseOver={[Function]}
                  onTouchEnd={[Function]}
                  onTouchStart={[Function]}
                >
                  <button
                    aria-label="Previous Page"
                    className="MuiButtonBase-root Mui-disabled MuiIconButton-root Mui-disabled MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
                    disabled={true}
                    onBlur={[Function]}
                    onClick={[Function]}
                    onContextMenu={[Function]}
                    onDragLeave={[Function]}
                    onFocus={[Function]}
                    onKeyDown={[Function]}
                    onKeyUp={[Function]}
                    onMouseDown={[Function]}
                    onMouseLeave={[Function]}
                    onMouseUp={[Function]}
                    onTouchEnd={[Function]}
                    onTouchMove={[Function]}
                    onTouchStart={[Function]}
                    tabIndex={-1}
                    type="button"
                  >
                    <svg
                      aria-hidden={true}
                      className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                      data-testid="chevron_left"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M15.41 7.41 14 6l-6 6 6 6 1.41-1.41L10.83 12z"
                      />
                    </svg>
                  </button>
                </span>
                <span
                  className="MuiTypography-root MuiTypography-caption css-1sn4lm3-MuiTypography-root"
                  style={
                    Object {
                      "alignSelf": "center",
                      "flex": 1,
                      "flexBasis": "inherit",
                      "textAlign": "center",
                    }
                  }
                >
                  0-0 of 0
                </span>
                <span
                  aria-label="Next Page"
                  aria-labelledby={null}
                  className=""
                  data-mui-internal-clone-element={true}
                  onBlur={[Function]}
                  onFocus={[Function]}
                  onMouseLeave={[Function]}
                  onMouseOver={[Function]}
                  onTouchEnd={[Function]}
                  onTouchStart={[Function]}
                >
                  <button
                    aria-label="Next Page"
                    className="MuiButtonBase-root Mui-disabled MuiIconButton-root Mui-disabled MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
                    disabled={true}
                    onBlur={[Function]}
                    onClick={[Function]}
                    onContextMenu={[Function]}
                    onDragLeave={[Function]}
                    onFocus={[Function]}
                    onKeyDown={[Function]}
                    onKeyUp={[Function]}
                    onMouseDown={[Function]}
                    onMouseLeave={[Function]}
                    onMouseUp={[Function]}
                    onTouchEnd={[Function]}
                    onTouchMove={[Function]}
                    onTouchStart={[Function]}
                    tabIndex={-1}
                    type="button"
                  >
                    <svg
                      aria-hidden={true}
                      className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                      data-testid="chevron_right"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M10 6 8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
                      />
                    </svg>
                  </button>
                </span>
                <span
                  aria-label="Last Page"
                  aria-labelledby={null}
                  className=""
                  data-mui-internal-clone-element={true}
                  onBlur={[Function]}
                  onFocus={[Function]}
                  onMouseLeave={[Function]}
                  onMouseOver={[Function]}
                  onTouchEnd={[Function]}
                  onTouchStart={[Function]}
                >
                  <button
                    aria-label="Last Page"
                    className="MuiButtonBase-root Mui-disabled MuiIconButton-root Mui-disabled MuiIconButton-sizeLarge css-mf1cb5-MuiButtonBase-root-MuiIconButton-root"
                    disabled={true}
                    onBlur={[Function]}
                    onClick={[Function]}
                    onContextMenu={[Function]}
                    onDragLeave={[Function]}
                    onFocus={[Function]}
                    onKeyDown={[Function]}
                    onKeyUp={[Function]}
                    onMouseDown={[Function]}
                    onMouseLeave={[Function]}
                    onMouseUp={[Function]}
                    onTouchEnd={[Function]}
                    onTouchMove={[Function]}
                    onTouchStart={[Function]}
                    tabIndex={-1}
                    type="button"
                  >
                    <svg
                      aria-hidden={true}
                      className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                      data-testid="last_page"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M5.59 7.41 10.18 12l-4.59 4.59L7 18l6-6-6-6zM16 6h2v12h-2z"
                      />
                    </svg>
                  </button>
                </span>
              </div>
            </div>
          </td>
        </tr>
      </tfoot>
    </table>
  </div>
</div>
`;
